pkg/deploy/controller/deploymentconfig/controller.go

```diff
@@ -277,3 +294,48 @@ func (c *DeploymentConfigController) updateStatus(config *deployapi.DeploymentCo
 	glog.V(4).Infof("Updated the status for %q (observed generation: %d)", deployutil.LabelForDeploymentConfig(config), config.Status.ObservedGeneration)
 	return nil
 }
+
+// cleanupOldDeployments triggers on deployment configuration update and add events in order
+// to delete old replication controller deployments if their quota has been reached
+func (c *DeploymentConfigController) cleanupOldDeployments(existingDeployments *kapi.ReplicationControllerList, deploymentConfig *deployapi.DeploymentConfig) error {
+	if deploymentConfig.Spec.RevisionHistoryLimit == nil {
+		// there is no past deplyoment quota set
+		return nil
+	}
+
+	activeDeployment := deployutil.ActiveDeployment(deploymentConfig, existingDeployments)
+
+	// we need to coerce the list type into something we can sort, and we need to filter out
+	// any deployments that we don't care about, namely the active deployment and any newer
+	// deployments
+	oldDeployments := []*kapi.ReplicationController{}
+	for i := 0; i < len(existingDeployments.Items); i++ {
+		if &existingDeployments.Items[i] == activeDeployment || deployutil.DeploymentVersionFor(&existingDeployments.Items[i]) > deployutil.DeploymentVersionFor(activeDeployment) {
+			continue
+		}
+		oldDeployments = append(oldDeployments, &existingDeployments.Items[i])
+	}
+
+	if len(oldDeployments) <= *deploymentConfig.Spec.RevisionHistoryLimit {
+		// the past deployment quota has not been exceeded
+		return nil
+	}
+
+	sort.Sort(controller.ControllersByCreationTimestamp(oldDeployments))
+
+	deletionErrors := []error{}
+	for i := 0; i < (len(oldDeployments) - *deploymentConfig.Spec.RevisionHistoryLimit); i++ {
+		deployment := oldDeployments[i]
+		if deployment.Spec.Replicas != 0 {
+			continue
+		}
+
+		err := c.kubeClient.ReplicationControllers(deployment.Namespace).Delete(deployment.Name)
```

Comment:
Note so that I won't forget it: you will need to remove deployer pods for failed deployments.